File: SupplierPortal.Web/Modules/ServerTypes/Market/RequestForm.ts
import { OfferDetailEditor } from "@/Market/OfferDetail/OfferDetailEditor";
import { LookupEditor, DateEditor, StringEditor, EnumEditor, DecimalEditor, PrefixedContext, initFormType } from "@serenity-is/corelib";
import { CurrencyType } from "../Modules/Market.Request.Enum.CurrencyType";

export interface RequestForm {
    CompanyName: LookupEditor;
    OfferDate: DateEditor;
    ValueDate: DateEditor;
    RequestState: StringEditor;
    CurrencyType: EnumEditor;
    Exchangerate: DecimalEditor;
    DeliveryTime: DateEditor;
    Representatives: StringEditor;
    ToWho: StringEditor;
    EMail: StringEditor;
    CallNumber: StringEditor;
    OfferId: OfferDetailEditor;
    SubTotal: DecimalEditor;
    LineDiscount: DecimalEditor;
    DiscountRate: DecimalEditor;
    DiscountTotal: DecimalEditor;
    GrossTotal: DecimalEditor;
    Vat: DecimalEditor;
    Total: DecimalEditor;
}

export class RequestForm extends PrefixedContext {
    static readonly formKey = 'Market.Request';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!RequestForm.init)  {
            RequestForm.init = true;

            var w0 = LookupEditor;
            var w1 = DateEditor;
            var w2 = StringEditor;
            var w3 = EnumEditor;
            var w4 = DecimalEditor;
            var w5 = OfferDetailEditor;

            initFormType(RequestForm, [
                'CompanyName', w0,
                'OfferDate', w1,
                'ValueDate', w1,
                'RequestState', w2,
                'CurrencyType', w3,
                'Exchangerate', w4,
                'DeliveryTime', w1,
                'Representatives', w2,
                'ToWho', w2,
                'EMail', w2,
                'CallNumber', w2,
                'OfferId', w5,
                'SubTotal', w4,
                'LineDiscount', w4,
                'DiscountRate', w4,
                'DiscountTotal', w4,
                'GrossTotal', w4,
                'Vat', w4,
                'Total', w4
            ]);
        }
    }
}

[CurrencyType]; // referenced types

File: SupplierPortal.Web/Modules/Market/Request/RequestForm.cs
using Serenity.ComponentModel;
using Serenity.Data;
using Serenity.Localization;
using SupplierPortal.Market.Enum;
using SupplierPortal.Modules.Market.Request.Enum;
using System;

namespace SupplierPortal.Market.Forms;

[FormScript("Market.Request")]
[BasedOnRow(typeof(RequestRow), CheckNames = true)]
public class RequestForm
{
    [LookupEditor(typeof(SupplierRow))]
    [QuarterWidth(UntilNext = true)]

    public string CompanyName { get; set; }
    //public int OfferId { get; set; }
    public DateTime OfferDate { get; set; }
    public DateTime ValueDate { get; set; }
    public State State { get; set; }
    public Currency Currency { get; set; }
    public CurrencyType CurrencyType { get; set; }
    public double Exchangerate { get; set; }
    public DateTime DeliveryTime { get; set; }
    public string Representatives { get; set; }
    public string ToWho { get; set; }
    public string EMail { get; set; }
    public string CallNumber { get; set; }

    [FullWidth]
    [DisplayName("Request"), RequestDetailEditor, IgnoreName]
    public List<RequestDetailRow> RequestId { get; set; }

    public decimal SubTotal { get; set; }
    [QuarterWidth]
    [CssClass("CalculatedFields")]
    [ReadOnly(true)]
    public decimal LineDiscount { get; set; }
    [QuarterWidth]
    [CssClass("CalculatedFields")]
    public decimal DiscountRate { get; set; }
    [QuarterWidth]
    [CssClass("CalculatedFields")]
    public decimal DiscountTotal { get; set; }
    [QuarterWidth]
    [CssClass("CalculatedFields")]
    [ReadOnly(true)]
    public decimal GrossTotal { get; set; }
    [QuarterWidth]
    [CssClass("CalculatedFields")]
    [ReadOnly(true)]
    public decimal Vat { get; set; }
    [QuarterWidth]
    [CssClass("CalculatedFields")]
    [ReadOnly(true)]
    public decimal Total { get; set; }
    //[Tab("Db.SalesMarketing.Sales.Detail")]
    //[Category("Müşteri Kodları"), Collapsible(true)]
}
